File: benchmark/nodejs/n_queens.js
const start = Date.now();
const board = new Array(8).fill(0);
let count = 0;
let row = 0;
let col = 0;

while (true) {
    if (row < 0) {
        break;
    }
    if (row === 8) {
        count++;
        row--;
        if (row >= 0) {
            col = board[row] + 1;
        }
    } else {
        if (col >= 8) {
            board[row] = 0;
            row--;
            if (row >= 0) {
                col = board[row] + 1;
            }
        } else {
            let safe = true;
            for (let i = 0; i < row; i++) {
                if (board[i] === col) {
                    safe = false;
                    i = row;
                } else {
                    let diff1 = board[i] - col;
                    if (diff1 < 0) { diff1 = -diff1; }
                    let diff2 = i - row;
                    if (diff2 < 0) { diff2 = -diff2; }
                    if (diff1 === diff2) {
                        safe = false;
                        i = row;
                    }
                }
            }
            if (safe) {
                board[row] = col;
                row++;
                col = 0;
            } else {
                col++;
            }
        }
    }
}
const end = Date.now();
console.log(`Solutions: ${count}`);
console.log(`Time: ${end - start}ms`);

File: benchmark/lua/n_queens.lua
local board = {}
for i = 0, 7 do board[i] = 0 end
local start = os.clock()
local count = 0
local row = 0
local col = 0

while true do
    if row < 0 then
        break
    end
    if row == 8 then
        count = count + 1
        row = row - 1
        if row >= 0 then
            col = board[row] + 1
        end
    else
        if col >= 8 then
            board[row] = 0
            row = row - 1
            if row >= 0 then
                col = board[row] + 1
            end
        else
            local safe = true
            local i = 0
            while i < row do
                if board[i] == col then
                    safe = false
                    i = row
                else
                    local diff1 = board[i] - col
                    if diff1 < 0 then
                        diff1 = 0 - diff1
                    end
                    local diff2 = i - row
                    if diff2 < 0 then
                        diff2 = 0 - diff2
                    end
                    if diff1 == diff2 then
                        safe = false
                        i = row
                    else
                        i = i + 1
                    end
                end
            end
            if safe then
                board[row] = col
                row = row + 1
                col = 0
            else
                col = col + 1
            end
        end
    end
end
local finish = os.clock()
print("Solutions: " .. count)
print(string.format("Time: %.0fms", (finish - start) * 1000))

